src/automata/ta_proto.cpp

```diff
@@ -0,0 +1,102 @@
+/***************************************************************************
+ *  ta_proto.cpp - Protobuf import/export for timed automata
+ *
+ *  Created:   Fri 19 Mar 10:50:08 CET 2021
+ *  Copyright  2021  Till Hofmann <hofmann@kbsg.rwth-aachen.de>
+ ****************************************************************************/
+/*  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Library General Public License for more details.
+ *
+ *  Read the full text in the LICENSE.md file.
+ */
+
+#include "automata/ta_proto.h"
+
+#include "automata/automata.h"
+#include "automata/ta.h"
+#include "automata/ta.pb.h"
+
+#include <algorithm>
+#include <range/v3/range/conversion.hpp>
+#include <range/v3/view/transform.hpp>
+#include <stdexcept>
+
+namespace automata::ta {
+
+namespace {
+Transition<std::string, std::string>
+parse_transition(const proto::TimedAutomaton::Transition &transition_proto)
+{
+	using ProtoClockConstraint = proto::TimedAutomaton::Transition::ClockConstraint;
+	std::multimap<std::string, ClockConstraint> clock_constraints;
+	for (const auto &clock_constraint : transition_proto.clock_constraints()) {
+		switch (clock_constraint.operand()) {
+		case ProtoClockConstraint::LESS:
+			clock_constraints.insert(
+			  {clock_constraint.clock(),
+			   AtomicClockConstraintT<std::less<Time>>{clock_constraint.comparand()}});
+			break;
+		case ProtoClockConstraint::LESS_EQUAL:
+			clock_constraints.insert(
+			  {clock_constraint.clock(),
+			   AtomicClockConstraintT<std::less_equal<Time>>{clock_constraint.comparand()}});
+			break;
+		case ProtoClockConstraint::EQUAL_TO:
+			clock_constraints.insert(
+			  {clock_constraint.clock(),
+			   AtomicClockConstraintT<std::equal_to<Time>>{clock_constraint.comparand()}});
+			break;
+		case ProtoClockConstraint::NOT_EQUAL_TO:
```

Comment:
Interesting: Apparently, here it is no problem to have a non-convex set. Thinking about it, this should be okay, since clocks are syntactically independent but it is an interesting observation, from a theoretical point of view.

Reply:
Addendum: I think this actually might be a problem, at least if someone tries to use the timed automata for anything else.
In our case we have to be carefully, since not equal constraints must at some point result in a case-split.

Reply:
Yes good point, I thought TAs have inequality, but apparently don't. I would not make this part of this PR, but remove inequality in a follow-up.

Reply:
Sounds legit, I mean right now it does not seem to be a problem, but I already see this biting us later.